Serialize/Deserialize source locations for instructions

usama54321 opened the pull request:
This PR serializes/deserializes source locations for instructions.

usama54321 committed:
Initial commit for serialization source locations

usama54321 commented:
@swift-ci test

usama54321 committed:
Skip unreachable code diagnostic if emitting for FilenameAndLocation
based SILLocation

usama54321 commented:
@swift-ci test

usama54321 commented:
@swift-ci test

usama54321 commented:
@swift-ci test

usama54321 committed:
Remove complete paths from test

usama54321 commented:
@swift-ci test

eeckstein commented on lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp:
```diff
@@ -994,6 +994,9 @@ static bool diagnoseUnreachableBlock(
     if (Loc.is<ImplicitReturnLocation>() || Loc.isAutoGenerated())
       return false;
 
+    if (Loc.isFilenameAndLocation())
```
Can you please add a comment why this is done?

usama54321 committed:
Handle different types of Source Locations. Remove redundant early
return from diagnoseUnreachableBlock

usama54321 commented:
@swift-ci test

usama54321 committed:
Serialize and deserialize implicit flag in SourceLocs. Also, skip
emitting unreachable block diagnostic for deserialized instructions as
it was suppressed already in the past as instructions were deserialized
as implicit by default

usama54321 commented:
@swift-ci test

adrian-prantl commented on include/swift/SIL/SILLocation.h:
```diff
@@ -552,6 +558,8 @@ class RegularLocation : public SILLocation {
       : SILLocation(L, RegularKind, Implicit) {}
   RegularLocation(FilenameAndLocation *filePos)
     : SILLocation(filePos, RegularKind) {}
+  RegularLocation(FilenameAndLocation *filePos, bool Implicit)
```
We should consider merging this with the function above and either update the call sites to explicitly state "explicit" or have a default argument.

adrian-prantl commented on include/swift/SIL/SILLocation.h:
```diff
@@ -622,6 +630,9 @@ class ReturnLocation : public SILLocation {
   ReturnLocation(FilenameAndLocation *filePos)
     : SILLocation(filePos, ReturnKind) {}
 
+  ReturnLocation(FilenameAndLocation *filePos, bool Implicit)
```
same here

adrian-prantl commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -394,6 +394,41 @@ SILDeserializer::readNextRecord(SmallVectorImpl<uint64_t> &scratch) {
+    FNameLoc = SILLocation::FilenameAndLocation::alloc(
+        Row, Col, MF->getIdentifierText(FNameID), SILMod);
+
+    ParsedLocs.insert({ParsedLocs.size() + 1, FNameLoc});
```
Is this a DenseMap of consecutive Int keys? If yes, why can't this a vector?

adrian-prantl commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -3074,6 +3083,47 @@ void SILSerializer::writeSILProperty(const SILProperty &prop) {
+    case SILLocation::ImplicitReturnKind:
+      LocationKind = SILLocation::ImplicitReturnKind;
+      break;
+    //TODO: handle other cases correctly
```
maybe explicitly list all the cases so we get a warning when a new one is added.

adrian-prantl reviewed (commented):
I only found a few minor comments.

eeckstein commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -394,6 +394,41 @@ SILDeserializer::readNextRecord(SmallVectorImpl<uint64_t> &scratch) {
+                                     SmallVectorImpl<uint64_t> &scratch) {
+  unsigned LocationKind = 0;
+  SILLocation::FilenameAndLocation *FNameLoc = nullptr;
+  if (kind == SIL_SOURCE_LOC_REF) {
```
Can you add a short comment explaining the strategy of serializing a sourceloc only the first time it appears and referencing this first sourceloc for successive appearances.

eeckstein reviewed (commented):
Can you please update the comment for `struct FilenameAndLocation`:
```
  /// In future we might also use it for de-serialized locations from a
  /// swiftmodule file.
```
The future is now 🙂

usama54321 committed:
Address feedback

usama54321 commented:
@swift-ci test

adrian-prantl commented on lib/SIL/IR/SILLocation.cpp:
```diff
@@ -137,6 +137,8 @@ SourceLoc SILLocation::getStartSourceLoc() const {
   if (isSILFile())
     return storage.sourceLoc;
+  if (getStorageKind() == FilenameAndLocationKind)
+    return SourceLoc();
```
This looks like it's working around incomplete serialization?

usama54321 committed:
Increment module format version

usama54321 committed:
Merge branch 'main' into usama54321/source-loc-serialization

usama54321 commented:
@swift-ci test
